### PR TITLE
I18n: Switch locale helpers

### PR DESCRIPTION
**Pull request by joallard**
This is as a follow-up to #1390 and #924.

> - Have the possibility to easily link to localized content in a language-agnostic way, ie. in a layout nav
> - Install a locale switcher link, which would link to the same resource, if available, in the target locale
> - Not relying on the locale dictionary to make this work

In practice, the following helpers are being added:
- [x] ~~`locale_path(page_id, locale = I18n.locale)`, so we can easily construct eg. a `<nav>` with language-agnostic keys.~~ (integrated into `url_for(*args, locale:)` and thus `link_to`)
- [x] `other_locales` and `other_locale`, some niceties for the following method...
- [ ] `switch_locale_path(locale = other_locale)`, so we can easily switch language while staying on the same resource

Internally,

I recycled `@maps` that seemed to be useless and used it as a localization map, it would now look like this:

``` rb
{"president"=>{:en=>"images/president.svg", :es=>"es/images/president.svg"},
 "hello"=>{:en=>"hello.html", :es=>"es/hola.html"},
 "flag"=>{:en=>"images/flag.svg", :es=>"es/images/flag.svg"},
 "index"=>{:en=>"index.html", :es=>"es/index.html"},
 "morning"=>{:es=>"es/manana.html", :en=>"morning.html"},
 "one"=>{:en=>"one.html", :es=>"es/una.html"},
 "partials.index"=>{:en=>"partials/index.html", :es=>"es/partials/index.html"}}
```

Which makes it easy to find a localized path.

The extension is in bad need of a refactor. I can do some things if the feedback is positive.

I've had to prioritize loading the Frontmatter extension before I18n, because we're using the `page_id` frontmatter key. Since Routing depends on I18n, this means Frontmatter is now before Routing as opposed to after. Don't know what this means for the rest of the features.

```
Failing Scenarios:
cucumber features/frontmatter_page_settings.feature:35 # Scenario: Overriding layout through frontmatter
```

Work in progress; please comment.

### CODE REVIEW

**Comment by tdreyno** [comment]
Thanks for diving in @joallard. I'm traveling right now, but will try to review and comment soon.

**Comment by tdreyno** [comment]
You're absolutely right, this extension is really crufty. I've merged in a lot of features, not knowing which are good or bad because I only use i18n in the simplest way, myself.

We'd love any direction, code and ideas you have.

I really like this iterative approach, slowly adding features rather than a complete do-over.

I'm going to push out v4 beta 1 today, so let's save this for beta 2. And we need to track down that failing test.

**Comment by joallard** [comment]
Awesome. I'll go forward then!

**Comment by joallard** [comment]
I can't find where the layout option has been implemented. The tests have been added after the implementation. Pointers?

**Comment by tdreyno** [comment]
Should be this: https://github.com/middleman/middleman/blob/master/middleman-core/lib/middleman-core/core_extensions/front_matter.rb#L44-L45

I think it's unrelated to i18n, but it is related to changing the order things run in.

**Comment by joallard** [comment]
About that failing test: it's because Routing overrides the Frontmatter layout setting with what's in `config.rb`. I changed the order to `{Routing: 60, FrontMatter: 70, I18n: 80}`, but `features/capture_html.feature` started to fail. I don't know what to do.

**Comment by joallard** [comment]
Can I get help on this?

**Comment by tdreyno** [comment]
Haven't had time. `capture_html` and all the Padrino internals are super gnarly and very prone to breakage. Lets change it so `capture_html` fails and we'll fix it post-merge.

**Comment by tdreyno** [comment]
@joallard Can this be brought up to date with beta2?

**Comment by joallard** [comment]
Okay, let me get to it

**Comment by tdreyno** [comment]
@joallard While you're at it, can you target the `v3-stable` branch?

**Comment by joallard** [comment]
Got it!

**Comment by joallard** [comment]
Can't do `v3-stable`. The diff is too large (it's a v4 extension and I was based on master—plus some differences in how the resources are fetched I think) and I don't understand what's there. I lost the context I had in Feb.

I rebased on master, but there are still some failures.

**Comment by Lightheaded** [comment]
Any updates on this issue?

**Comment by Lightheaded** [comment]
Any updates for the issue?

**Comment by joallard** [comment]
@tdreyno Is it okay if this is a master-only patch? (ie v4 I presume?)

**Comment by tdreyno** [comment]
Yes

File: middleman-core/lib/middleman-core/core_extensions/i18n.rb [hunk 1]
```diff
@@ -89,6 +92,31 @@ def locate_partial(partial_name, try_static=false)
...
+
+    # Path to the same page in another locale
+    def switch_locale_path(locale=other_locale)
+      raise ArgumentError, 'No other locale to switch to' unless locale
+      locale_path(current_page.locals[:page_id], locale)
+    end
+
+    def other_locale
+      other_locales[0]
+    end
```
Comment (johnnyshields) [inline middleman-core/lib/middleman-core/core_extensions/i18n.rb]: This method seems superfluous, is it really needed?

**Comment by johnnyshields** [comment]
Does this PR add support for I18n'ed frontmatter keys (as per the title of the PR)? I don't see it in the code...

**Comment by joallard** [comment]
@tdreyno I'm a little busy at the moment, and my confusion with the failing tests and ext load order (among larger components) has prevented me from being more proactive in fixing this PR, but let me know if there's anything I can do or answer. I'd be happy to help Middleman get a killer i18n any way I can, it's a truly great project you've made you guys.

**Comment by tdreyno** [comment]
Cool. I'll try rebasing and see where things stand now. Thanks!

**Comment by pixelchef** [comment]
Would really love to use this as a clean way to link to locale blog.sources in config.rb to keep articles from different locales separate!

**Comment by joallard** [comment]
9de1f16f

**Comment by joallard** [comment]
Seeing `url_for` has been implemented in parallel, here are the parts that this PR is aiming to:

1. I can link to the same page in another locale (⇒ 2)
2. I can link to another page in an arbitrary locale (⇒ 3)
3. Every page has a `page_id` that is unique, rememberable and independent of locale
4. I can change the `page_id` in a page frontmatter

**Comment by joallard** [comment]
What I'm tiptoeing around is the `page_id` logic. That mechanism isn't really solid, and it seems to me the urls in general should be able to refer to a certain `page_id`, a little à la how Rails has `as` labels for routing. 

(Btw, the rebase fixed the extension priorities that seem to have magically been sorted out)

**Comment by joallard** [comment]
@tdreyno, what's your take on this?

File: middleman-core/lib/middleman-core/core_extensions/i18n.rb [hunk 2]
```diff
@@ -256,10 +273,20 @@ def parse_locale_extension(path)
     [locale, path, basename]
   end
 
-  LocalizedPageDescriptor = Struct.new(:path, :source_path, :locale) do
+  # Turn path into a page_id for the lookup map
+  # eg. "foo/bar/baz.foo.html" => "foo.bar.baz.foo"
```
Comment (tdreyno) [inline middleman-core/lib/middleman-core/core_extensions/i18n.rb]: Why is the actual path string not enough? What does changing this into all dots get us? Is it intended to be an actual structure from a locale file?

**Comment by tdreyno** [comment]
I link an arbitrary `page_id` concept or linking in MM is a really good idea. Would the `page_id` be the same for all translations of the same resource?

That said, this is looking pretty good. Those helpers will be nice.

File: middleman-core/lib/middleman-core/core_extensions/i18n.rb (hunk 2 above)
Comment (joallard) [inline middleman-core/lib/middleman-core/core_extensions/i18n.rb]: The idea is that it is then easy to use it as a key for lookup into I18n. So one could do: `{foo: {bar: {baz: {foo: "Text"}}}}` However thinking about it we could probably keep the slashes for `page_id` and do the conversion into I18n lookup key later in the process. But that's the idea behind.

**Comment by joallard** [comment]
> Would the page_id be the same for all translations of the same resource?

Exactly. Why it is there is that for localized versions of the same file, eg. `about.en.html` and `a-propos.fr.html`, the filename needs to be different to reflect the url, even if it represents the same resource (in a different language format). The default is to take the filename/path, but we need to adjust that in the other languages so they can be "linked" together. Hence one could put `page_id: about` in the frontmatter of `a-propos.fr.html` and be all set.

I just thought it's a little clunky to tack that logic just in I18n while it could be baked in deeper into Middleman as it answers to how are pages identified, regardless of their language. We could then say let's `link_to t(:about), :about` instead of `link_to t(:about), "/about.html"`.

**Comment by tdreyno** [comment]
@joallard I agree, let's put that into core. It can be a piece of general metadata on a resource, set by `page` or `proxy` and overridden by `frontmatter`. Let me take a crack at that today.

**Comment by tdreyno** [comment]
Okay @joallard, how's this for a nice base: https://github.com/middleman/middleman/commit/edb1a53060abdc3589cbcfb6a60b69796462c999

**Comment by joallard** [comment]
Yep, I'm on it

**Comment by johnnyshields** [comment]
@joallard @tdreyno what's the latest on this one?

**Comment by joallard** [comment]
Rebasing was a little less straightforward than I thought and I ran out of time, but it's on my to-do list, between real life things and maintaining Selectize. I can share my WIP if that helps.

**Comment by tdreyno** [comment]
@joallard WIP is always going, I'm going on vacation next week, I may actually have to to investigate this

**Comment by joallard** [comment]
@tdreyno Not sure I understood this last one

**Comment by johnnyshields** [comment]
@joallard maybe it might be easier to break this into smaller tickets that can be merged one-by-one?

**Comment by joallard** [comment]
Whew! And here we are. Works. Rebased. As we can see (I activated verbose commit mode), it pretty much comes in one piece. It's dependent on #1996, so we'll solve that first.

What's left:
- [ ] That `switch_locale_path` helper
- [ ] Clean up `url_for`
- [ ] Clean up Git history

We'll probably get some edge cases too to add in there.

**Comment by joallard** [comment]
Also, is there a reason we don't use Pry out of the box in dev?

**Comment by joallard** [comment]
Next up: remove the (*) `localizable` folder.

**Comment by tdreyno** [comment]
@joallard I think because we used to support versions of Ruby which didn't support pry... but now, good to include by default

**Comment by manuelmeurer** [comment]
Hi @joallard and @tdreyno, what it the status of this PR? I just ran into the issue that `link_to`/`url_for` with I18n doesn't work as expected (at least with `directory_indexes`, I didn't check without). Will the changes in this PR fix this? What's needed to move forward?

**Comment by johnnyshields** [comment]
FYI here's my I18n helper methods I've added to my Middleman setup: https://gist.github.com/johnnyshields/98a695df51b1e99f3593579d3c9a3fd1

One thing to note is that my primary locale is Japanese, so my site root (/) is Japanese, and other locales are at /en/... etc. Would be good if this can be support in this PR.

**Comment by joallard** [comment]
I had a refactor lingering, it's pushed at 6e89a36. @tdreyno you'll probably notice that I've got references to the `i18n` in my classes everywhere, should this have been structs?

**Comment by jnf-si** [comment]
Hi! I just wanted to take a moment to thank yinz for the ongoing work on this issue. This (year-long!) thread has been really informative and has helped me figure out how to move forward with my Middleman project. I also really appreciate the supportive tone and graceful acknowledgement that OSS is hard, language is hard, and people are busy.

Thanks again for sticking with this. 👍

**Comment by tdreyno** [comment]
Looking for a passing build, but happy to review whenever

**Comment by joallard** [comment]
I don't have so much time to dedicate to this right now, but it should be only a matter of fixing a few loose ends on the tip of the branch. I'm happy to help if someone wants to pick it back up.

**Comment by tdreyno** [comment]
That liquid error is a whole other thing I need to clean up (likely by removing tests for old versions of liquid), you can ignore it

**Comment by kstratis** [comment]
Hi, any update on this one?

**Comment by dorellang** [comment]
I've been hacking Middleman a lot (by monkey patching and reopening some framework classes) to get Internationalization, Blogging and Directory Indexes playing along for my personal website. I think I grasp how the code currently on master works and I'd love to contribute, but I don't know what the project standards are, how the code should be tested, if there's any backwards compatible behavior that should not be broken, etc.

I'd really appreciate any information. I'm looking forward to solving this issue, since the solution I've made for my website just feels like a complete duct tape mess and I'm pretty sure that would break in the next Middleman release. It just would be freaking nice to get it working in a cleaner way.

Cheers!

**Comment by ChristianPeters** [comment]
I also spent a lot of time, having to understand nearly all messy internals to be able to duct tape our locale switch helpers together.
Sadly, I can't find any time to contribute but I could share my helper structure if that might be helpful.

**Comment by dagumak** [comment]
Any updates?

**Comment by tdreyno** [comment]
@dorellang If you can write a specification of how it should work, we can convert that to a Cucumber test

**Comment by joallard** [comment]
This was a good run, y'all, it brings back memories. 😉 Glad you were able to clean this one up @tdreyno

**Pull request closed without merging**